### PR TITLE
Umap

### PR DESCRIPTION
**Pull request by csweaver**
Add Umap coords as an option for graph layout.

### CODE REVIEW

File: application.py [hunk 1]
```diff
@@ -35,17 +35,21 @@ class QueryStringError(Exception):
...
+        self.z_coord = z
 
     def x(self):
         return float(self.x_coord)
 
     def y(self):
         return float(self.y_coord)
 
+    def z(self):
+        return float(self.z_coord)
```
Comment (ttung) [inline application.py]: is it ok that `z_coord` is None when this is called? (i.e., either you guarantee that you'll never call z() when `z_coord` is `None`, or the exception is handled)

File: application.py [hunk 2]
```diff
@@ -613,10 +617,14 @@ def create_graph(output_cellset, includeisolated=False, graphname="cellgraph"):
     """
     global e
```
Comment (ttung) [inline application.py]: unrelated to this diff: what's this global?

File: application.py [hunk 3]
```diff
@@ -613,10 +617,14 @@ def create_graph(output_cellset, includeisolated=False, graphname="cellgraph"):
...
     # Graph using EM2 or use metadata coordinates
-    if application.config["GRAPH_EM2"]:
+    if application.config["GRAPH_METHOD"] == "em2":
```
Comment (ttung) [inline application.py]: if this becomes a common pattern, you can use an enum to store the name of the graph method along with the function call.

File: application.py [hunk 4]
```diff
@@ -613,10 +617,14 @@ def create_graph(output_cellset, includeisolated=False, graphname="cellgraph"):
...
         graph, cellidlist = create_graph_em2(output_cellset, includeisolated=False, graphname="cellgraph")
-    else:
+    elif application.config["GRAPH_METHOD"] == "umap":
+        graph, cellidlist = create_graph_umap(output_cellset)
+    elif application.config["GRAPH_METHOD"] == "tsne":
         graph, cellidlist = create_graph_tsne(output_cellset)
+    else:
+        raise ValueError("Invalid graph method :{}".format(application.config["GRAPH_METHOD"]))
```
Comment (ttung) [inline application.py]: is this py36+?  if so, you can use f-style strings!

File: application.py [hunk 5]
```diff
@@ -661,6 +669,26 @@ def create_graph_tsne(output_cellset):
...
+    """
+    # get metadata for each cell
+    cellidlist = e.getCellSet(output_cellset)
+    vertices = [Vertex(get_cell_name(cid),
+                       e.getCellMetaDataValue(cid, "umap_coord1"),
+                       e.getCellMetaDataValue(cid, "umap_coord2"),
+                       e.getCellMetaDataValue(cid, "umap_coord3")) for cid in cellidlist]
+    normalized_verticies = normalize_verticies(vertices)
+    graph = []
+    for i in range(len(normalized_verticies["labels"])):
```
Comment (ttung) [inline application.py]: possible alternative:

```
graph = [
    (label, x, y)
    for label, x, y in zip(normalized_vertices['labels'], normalized_vertices['x'], normalized_vertices['y'])
]
```

File: application.py (hunk 3 above)
Comment (csweaver) [reply application.py]: Good idea. I'll add that into my epic on being able to plug in various methods for the computations

File: application.py (hunk 4 above)
Comment (csweaver) [reply application.py]: Wow! I knew javascript had these, I didn't realize python had gotten onboard

Comment (csweaver) [reply application.py]: Looks like I'm on 3.5

File: application.py (hunk 2 above)
Comment (csweaver) [reply application.py]: Part of how EM2 works... Working on refactoring it

File: application.py (hunk 1 above)
Comment (csweaver) [reply application.py]: Yep, I'll handle it when we get a 3D visualization option

**Pull request merged**